litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go

```diff
@@ -2,42 +2,124 @@ package k8s
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"log"
 	"os"
+	"strings"
+	"sync"
+	"time"
 
 	yaml_converter "github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+	yaml2 "gopkg.in/yaml.v2"
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	memory "k8s.io/client-go/discovery/cached"
 	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/restmapper"
-
-	memory "k8s.io/client-go/discovery/cached"
 )
 
 const (
 	ExternAgentConfigName = "agent-config"
+	LiveCheckMaxTries     = 6
 )
 
+type AgentComponents struct {
+	Deployments      []string `yaml:"DEPLOYMENTS"`
+	LiveStatus       bool
+	AccessLiveStatus sync.Mutex
+}
+
 var (
 	Ctx             = context.Background()
 	decUnstructured = yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
 	dr              dynamic.ResourceInterface
 	AgentNamespace  = os.Getenv("AGENT_NAMESPACE")
 )
 
+func CheckComponentStatus(componentEnv string) error {
+	if componentEnv == "" {
+		return errors.New("components not found in agent config")
+	}
+
+	clientset, err := GetGenericK8sClient()
+	if err != nil {
+		return err
+	}
+
+	var components AgentComponents
+
+	err = yaml2.Unmarshal([]byte(strings.TrimSpace(componentEnv)), &components)
+	if err != nil {
+		return err
+	}
+
+	components.LiveStatus = true
+	components.AccessLiveStatus = sync.Mutex{}
+
+	wait := sync.WaitGroup{}
+
+	// add all agent components to waitgroup
+	wait.Add(1)
+	go checkDeploymentStatus(&components, clientset, &wait)
+
+	wait.Wait()
+	if !components.LiveStatus {
+		return errors.New("all components failed to startup")
+	}
+	return nil
+}
+
+func checkDeploymentStatus(components *AgentComponents, clientset *kubernetes.Clientset, wait *sync.WaitGroup) {
+	downCount := 0
+	retries := 0
+	defer wait.Done()
+	for retries < LiveCheckMaxTries {
+		for _, dep := range components.Deployments {
+			depSpec, err := clientset.AppsV1().Deployments(AgentNamespace).Get(dep, metav1.GetOptions{})
+			if err != nil {
+				logrus.Errorf("failed to get deployments %v , err : %v", dep, err.Error())
+				downCount += 1
+				continue
+			}
+			if depSpec == nil {
+				logrus.Errorf("failed to get deployments %v , err : deployment not found", dep)
+				downCount += 1
+				continue
+			}
+			if depSpec.Status.Replicas != depSpec.Status.AvailableReplicas {
```

Comment:
The availableReplicas can be a bit misleading ? For example - the case where the deployments are crashing _after_ running (due to some issue, say a missing CRD), the available Replica count can fluctuate (1 or more when running and 0 when not). It might be safer to check the pod phase and containerStatuseses (reason will show up) - as we do in case of experiments

Reply:
According to the doc for AvailableReplica - `Total number of available pods (ready for at least minReadySeconds) targeted by this deployment` I think this will only count the pods that are live and running for the specified amount of time. In case it crashes after a successful run I think the pod phase will also not be helpful because at the point when the subscriber is checking the phase could be running too.
Also we can add a custom minReadySeconds in our deployments to make this more resilient

Reply:
Have updated it to use container status